internal/provider/resource_credential.go

```diff
@@ -217,44 +315,118 @@ func resourceCredentialUpdate(ctx context.Context, d *schema.ResourceData, meta
 		Name:                 credentialName,
 	})
 	if err != nil {
-		return diag.FromErr(err)
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to update credential resource, got error: %s", err))
+		return
 	}
+	tflog.Trace(ctx, fmt.Sprintf("updated credential resource %q", credentialName))
 
-	newID := fmt.Sprintf("%s:%s:%t:%t", credentialName, cloudName, newClientCredential, newControllerCredential)
-	d.SetId(newID)
+	data.ID = types.StringValue(newIDFrom(credentialName, cloudName, newClientCredential, newControllerCredential))
 
-	return diags
+	// Write the updated state data into the Response.State
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
 }
 
-func resourceCredentialDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	// When removing cloud credential from a controller, Juju performs additional
-	// checks to ensure that there are no models using this credential. The provider will not force the removal
-	client := meta.(*juju.Client)
-	var diags diag.Diagnostics
+func (c *credentialResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	// Check first if the client is configured
+	if c.client == nil {
+		addClientNotConfigured(&resp.Diagnostics, "delete")
+		return
+	}
+
+	var data *credentialResourceModel
+
+	// Read Terraform configuration from the request into the resource model
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Access prior state data
 
-	id := strings.Split(d.Id(), ":")
-	if len(id) != 4 {
-		return diag.Errorf("unable to parse credential name and cloud name from provided ID")
+	resID := retrieveValidateID(data, &resp.Diagnostics, "update")
+	if resp.Diagnostics.HasError() {
+		return
 	}
-	credentialName, cloudName, clientCredentialStr, controllerCredentialStr := id[0], id[1], id[2], id[3]
+	// extract : name & cloud.name, client_credential, controller_credential
+	credentialName, cloudName, clientCredentialStr, controllerCredentialStr := resID[0], resID[1], resID[2], resID[3]
 	clientCredential, controllerCredential, err := convertOptionsBool(clientCredentialStr, controllerCredentialStr)
```

Comment:
can we embed this in retreiveValidateID?

func retrieveValidateCredentialID(id types.String,  diags *diag.Diagnostics, method string) (string, string, bool, bool)